tensorly/decomposition/parafac2.py

```diff
@@ -0,0 +1,280 @@
+        rhs = T.transpose(tensor_slice)
+        U, S, Vh = svd_fun(T.dot(lhs, rhs), n_eigenvecs=A.shape[1])
+
+        out[i] = tl.index_update(projection, tl.index[:], T.transpose(T.dot(U, Vh)))
+
+    return out
+
+
+def _project_tensor_slices(tensor_slices, projections, out=None):
+    if out is None:
+        rank = projections[0].shape[1]
+        num_slices = len(tensor_slices)
+        num_cols = tensor_slices[0].shape[1]
+        out = T.zeros((num_slices, rank, num_cols), **T.context(tensor_slices[0]))
+
+    for i, (tensor_slice, projection) in enumerate(zip(tensor_slices, projections)):
+        slice_ = T.dot(T.transpose(projection), tensor_slice)
+        out = tl.index_update(out, tl.index[i, :], slice_)
+    return out
+
+
+def _get_svd(svd):
+    if svd in tl.SVD_FUNS:
+        return tl.SVD_FUNS[svd]
+    else:
+        message = 'Got svd={}. However, for the current backend ({}), the possible choices are {}'.format(
+                svd, tl.get_backend(), tl.SVD_FUNS)
+        raise ValueError(message)
+
+
+def _parafac2_reconstruction_error(tensor_slices, decomposition):
+    _validate_parafac2_tensor(decomposition)
+    squared_error = 0
+    for idx, tensor_slice in enumerate(tensor_slices):
+        reconstruction = parafac2_to_slice(decomposition, idx, validate=False)
+        squared_error += tl.sum((tensor_slice - reconstruction)**2)
+    return tl.sqrt(squared_error)
+
+
+def parafac2(tensor_slices, rank, n_iter_max=100, init='random', svd='numpy_svd', normalize_factors=False,
+             tol=1e-8, random_state=None, verbose=False, return_errors=False, n_iter_parafac=5):
+    r"""PARAFAC2 decomposition [1]_ via alternating least squares (ALS)
+
+    Computes a rank-`rank` PARAFAC2 decomposition of the tensor defined by `tensor_slices`. 
+    The decomposition is on the form :math:`(A [B_i] C)` such that the i-th frontal slice,
+    :math:`X_i`, of :math:`X` is given by
+
+    .. math::
+    
+        X_i = B_i diag(a_i) C^T,
+    
+    where :math:`diag(a_i)` is the diagonal matrix whose nonzero entries are equal to
+    the :math:`i`-th row of the :math:`I \times R` factor matrix :math:`A`, :math:`B_i` 
+    is a :math:`J_i \times R` factor matrix such that the cross product matrix :math:`B_{i_1}^T B_{i_1}`
+    is constant for all :math:`i`, and :math:`C` is a :math:`K \times R` factor matrix. 
+    To compute this decomposition, we reformulate the expression for :math:`B_i` such that
+
+    .. math::
+
+        B_i = P_i B,
+
+    where :math:`P_i` is a :math:`J_i \times R` orthogonal matrix and :math:`B` is a
+    :math:`R \times R` matrix.
+
+    An alternative formulation of the PARAFAC2 decomposition is that the tensor element
+    :math:`X_{ijk}` is given by
+
+    .. math::
+
+        X_{ijk} = \sum_{r=1}^R A_{ir} B_{ijr} C_{kr},
+    
+    with the same constraints hold for :math:`B_i` as above.
+     
+
+    Parameters
+    ----------
+    tensor_slices : ndarray or list of ndarrays
```

Comment:
Maybe I can recommend that you further develop in the comments that 
- your code only works for 3d order tensors (unless I am mistaken). although the Parafac2 model exist in any order.
- the slices are "horizontal" (whereas from my experience, most people would intuitively work with "vectical" slices). This is to avoid that a user confuses the output factors A and C with each others.

Reply:
We updated the docstring to make this clearer.